src/test/unit/mcmc/chainset_test.cpp

```diff
@@ -0,0 +1,334 @@
+      chain_2.num_samples());
+
+  std::vector<stan::io::stan_csv> bernoulli;
+  bernoulli.push_back(bernoulli_default);
+  bernoulli.push_back(bernoulli_thin);
+  bernoulli.push_back(bernoulli_warmup);
+  stan::mcmc::chainset chain_3(bernoulli);
+  EXPECT_EQ(3, chain_3.num_chains());
+  EXPECT_EQ(bernoulli_default.header.size(), chain_3.num_params());
+  EXPECT_EQ(bernoulli_default.metadata.num_samples, chain_3.num_samples());
+}
+
+TEST_F(McmcChains, addFail) {
+  std::vector<stan::io::stan_csv> bad;
+  bad.push_back(bernoulli_default);
+  bad.push_back(bernoulli_500);
+  EXPECT_THROW(stan::mcmc::chainset fail(bad), std::invalid_argument);
+
+  bad.clear();
+  bad.push_back(bernoulli_default);
+  bad.push_back(eight_schools_1);
+  EXPECT_THROW(stan::mcmc::chainset fail(bad), std::invalid_argument);
+
+  bad.clear();
+  bad.push_back(bernoulli_default);
+  bad.push_back(bernoulli_zeta);
+  EXPECT_THROW(stan::mcmc::chainset fail(bad), std::invalid_argument);
+}
+
+TEST_F(McmcChains, paramNameIndex) {
+  stan::mcmc::chainset chains_csv(eight_schools_1);
+  EXPECT_EQ(1, chains_csv.num_chains());
+  for (int i = 0; i < eight_schools_1.header.size(); i++) {
+    EXPECT_EQ(eight_schools_1.header[i], chains_csv.param_name(i));
+    EXPECT_EQ(i, chains_csv.index(eight_schools_1.header[i]));
+  }
+  EXPECT_THROW(chains_csv.param_name(-5000), std::invalid_argument);
+  EXPECT_THROW(chains_csv.param_name(5000), std::invalid_argument);
+  EXPECT_THROW(chains_csv.index("foo"), std::invalid_argument);
+}
+
+TEST_F(McmcChains, eight_schools_samples) {
+  std::vector<stan::io::stan_csv> eight_schools;
+  eight_schools.push_back(eight_schools_1);
+  eight_schools.push_back(eight_schools_2);
+  stan::mcmc::chainset chain_2(eight_schools);
+  Eigen::MatrixXd mu_all = chain_2.samples("mu");
+  EXPECT_EQ(chain_2.num_chains() * chain_2.num_samples(), mu_all.size());
+  mu_all = chain_2.samples(7);
+  EXPECT_EQ(chain_2.num_chains() * chain_2.num_samples(), mu_all.size());
+
+  EXPECT_THROW(chain_2.samples(5000), std::invalid_argument);
+  EXPECT_THROW(chain_2.samples("foo"), std::invalid_argument);
+}
+
+TEST_F(McmcChains, split_rank_normalized_rhat) {
+  stan::mcmc::chainset chain_1(eight_schools_1);
+  EXPECT_EQ(1, chain_1.num_chains());
+
+  // test against R implementation in pkg posterior
+  Eigen::VectorXd rhat_8_schools_1_bulk(10);
+  rhat_8_schools_1_bulk << 1.0012958313, 1.0046136496, 1.0085723580,
+      1.0248629375, 1.0111456620, 1.0004458336, 0.9987162973, 1.0339773469,
+      0.9985612618, 1.0281667351;
+
+  Eigen::VectorXd rhat_8_schools_1_tail(10);
+  rhat_8_schools_1_tail << 1.005676523, 1.009670999, 1.00184184, 1.002222679,
+      1.004148161, 1.003218528, 1.009195353, 1.001426744, 1.003984381,
+      1.025817745;
+
+  for (size_t i = 0; i < 10; ++i) {
+    auto rhats = chain_1.split_rank_normalized_rhat(i + 7);
+    EXPECT_NEAR(rhats.first, rhat_8_schools_1_bulk(i), 0.05);
+    EXPECT_NEAR(rhats.second, rhat_8_schools_1_tail(i), 0.05);
```

Comment:
Having an absolute tolerance of 0.05 completely invalidates this test. At that level of tolerance, both of these rhat vectors are indistinguishable from a vector of 1s

I think the number of digits in the expected vectors is reasonable, so it would be great if this test passed with the tolerance more like `1e-8`, or even `1e-10`, which it seems like most of the other uses of EXPECT_NEAR in the code base use

Reply:
this isn't possible - the compound differences introduced between the  C++ implementation of rank-normalization, and computations of the between and within chain variances and the corresponding operations in R result in values that have at least 0.03 difference for these tests.

Reply:
If that is the case, we shouldn't test against R at all, and instead compute some values to the desired precision and "bless" them as correct. It requires a lot more confidence in the existing C++ implementation, though, so I'd try to track down if we couldn't have C++ and R agree better first